### PR TITLE
Creating module for HTTP client with Retry approach to deal with network hiccups

### PR DESCRIPTION
**Pull request by rd4398**
Fixes: https://github.com/python-wheel-build/fromager/issues/222

Let me know if this looks ok. I will then proceed to make similar changes related to `requests.get()` that I encountered in `resolver.py`

### CODE REVIEW

**Comment by tiran** [comment]
Nice work, @rd4398 

I would design this slightly different, so we have one central place to control, configure, and reuse a request session. A common session object also allows for session resumption and HTTP pipelining. The `WorkContext` object would be a logical place. It already holds configuration and state information.

```
class WorkContext:
    def __init__(...):
        ...
        self.requests = self._make_requests_session()
        ...

    def _make_requests_session(self) -> requests.Session:
        ...
```

then pass WorkContext to `download_url` and change `requests.get(...)`  to `ctx.requests.get`. This gives us one place to control download parameter.

How does `Retry` work? Does it resume a file download when it gets interrupted in the middle of an operation?

File: src/fromager/http_client.py [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+It has been adopted from: https://www.raelldottin.com/2023/09/enhancing-resilience-in-your-python.html
+"""
+
+import requests
+from requests.adapters import HTTPAdapter, Retry
+
+
+class HTTPClient:
+    # Set up the retries, backoff_factor and status_forcelist
+    def __init__(self, base_url, retries=5, backoff_factor=0.1, status_forcelist=None):
```
Comment (dhellmann) [inline src/fromager/http_client.py:12]: If base_url isn't used until the call to get(), does it need to be an argument here?

File: src/fromager/http_client.py [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+            status_forcelist=status_forcelist,
+        )
+
+        # Initialize the adapter object
+        adapter = HTTPAdapter(max_retries=retries)
+        self.session.mount("http://", adapter)
+        self.session.mount("https://", adapter)
+
+    # Get request function to make HTTP GET request using session and Retry
+    def get_request(self, endpoint="", **kwargs):
```
Comment (dhellmann) [inline src/fromager/http_client.py:32]: Maybe call this just get() so it mirrors the requests library API?

File: src/fromager/sources.py [hunk 3]
```diff
@@ -199,12 +200,16 @@ def download_url(
...
-        r.raise_for_status()
-        with open(outfile, "wb") as f:
-            logger.debug(f"writing to {outfile}")
-            for chunk in r.iter_content(chunk_size=1024 * 1024):
-                f.write(chunk)
+
+    # Initialize the HTTPClient module
+    client = HTTPClient(url)
+    # Call HTTP GET for the URL
+    r = client.get_request()
```
Comment (dhellmann) [inline src/fromager/sources.py:207]: We lost the `stream=True` argument there, which is important for making the loop below work. Without that argument requests downloads the whole file in one go, which means you have to be able to hold the whole thing in RAM. We're downloading very large wheels in some case, so we don't want to allocate that much RAM at one time.

**Comment by rd4398** [comment]
> Nice work, @rd4398
> 
> I would design this slightly different, so we have one central place to control, configure, and reuse a request session. A common session object also allows for session resumption and HTTP pipelining. The `WorkContext` object would be a logical place. It already holds configuration and state information.
> 
> ```
> class WorkContext:
>     def __init__(...):
>         ...
>         self.requests = self._make_requests_session()
>         ...
> 
>     def _make_requests_session(self) -> requests.Session:
>         ...
> ```
> 
> then pass WorkContext to `download_url` and change `requests.get(...)` to `ctx.requests.get`. This gives us one place to control download parameter.
> 
> How does `Retry` work? Does it resume a file download when it gets interrupted in the middle of an operation?

No, it doesn't resume the download. It simply retries based on the `backoff_factor`. If the` backoff_factor = 0.1`, it will sleep for ` [0.05s, 0.1s, 0.2s, 0.4s, ...] between retries. ` depending on the `max_retries` parameter

File: src/fromager/sources.py (hunk 3 above)
Comment (rd4398) [reply src/fromager/sources.py:207]: Okay, I will make the changes and try to design this in a way @tiran suggested above

**Comment by tiran** [comment]
> No, it doesn't resume the download. It simply retries based on the `backoff_factor`. If the` backoff_factor = 0.1`, it will sleep for `[0.05s, 0.1s, 0.2s, 0.4s, ...] between retries.` depending on the `max_retries` parameter

How is Retry going to help us to resume downloads? Either Retry helper must resume at exactly the same offset or we need to recognize resumption somehow and start download + write to file from scratch.

**Comment by rd4398** [comment]
I made changes in the design as per suggestion above. Does it look ok? Should I proceed to make similar changes for other `request.get()` in `resolver.py`?

**Comment by rd4398** [comment]
> > No, it doesn't resume the download. It simply retries based on the `backoff_factor`. If the` backoff_factor = 0.1`, it will sleep for `[0.05s, 0.1s, 0.2s, 0.4s, ...] between retries.` depending on the `max_retries` parameter
> 
> How is Retry going to help us to resume downloads? Either Retry helper must resume at exactly the same offset or we need to recognize resumption somehow and start download + write to file from scratch.

I researched about it but could not find anything. Even the documentation for [Retry](https://urllib3.readthedocs.io/en/latest/reference/urllib3.util.html#module-urllib3.util.retry) doesn't mention about it. 
Do you have any idea how we can implement this? 
(Also, not sure whether this should be included in the current PR or a new issue needs to be filed)

**Comment by rd4398** [comment]
code has changed a lot since this PR which resulted in a messy rebase. Thus closing this PR.

**Pull request closed without merging**